### PR TITLE
completions/ssh: Include global ssh hosts

### PR DESCRIPTION
**Pull request by plasmastorm**
Partial fix for https://github.com/ohmybash/oh-my-bash/issues/568

### CODE REVIEW

**Review by akinomyoga — COMMENTED** [review]
Thank you for your contribution. I merged another PR of ssh_config #529. The suggested addition of the support for `/etc/ssh/ssh_{config,known_hosts}` is good. However, I'd like to reduce the code duplicates before merging.

Could you rebase on top of the latest `master` and modify the code to store the found files in an array and run e.g. `grep ^HOST "${config_files[@]}"` at once?

```bash
local -a config_files=()
if [[ -r ~/.ssh/config ]]; then
  config_files+=(~/.ssh/config)
  ...
fi
if [[ -r /etc/ssh/ssh_config ]]; then
  config_files+=(/etc/ssh/ssh_config)
fi
COMPREPLY+=($(compgen -W "$(awk '/^Host/ {for (i=2; i<=NF; i++) print $i}' "${config_files[@]}")" "${options[@]}"))
```

Similar for known_hosts.

**Comment by plasmastorm** [comment]
Will do - `/etc/ssh/ssh_config` can also include files, worth trying to work it in with the new changes by looping over both config files?

**Comment by akinomyoga** [comment]
Ah, right. I think the Include there can also be processed.

File: completions/ssh.completion.sh [hunk 1]
```diff
@@ -16,37 +16,42 @@ function _omb_completion_ssh {
...
-    _omb_util_glob_expand include_files '${include_patterns[*]}'
-    local include_file
-    for include_file in "${include_files[@]}";do
-      # parse all defined hosts from that file
-      [[ -s $include_file ]] && config_files+=("$include_file")
-    done
-
-    COMPREPLY+=($(compgen -W "$(awk '/^Host/ {for (i=2; i<=NF; i++) print $i}' "${config_files[@]}")" "${options[@]}"))
-  fi
+  for base_config_file in ~/.ssh/config /etc/ssh/ssh_config; do
```
Comment (akinomyoga) [inline completions/ssh.completion.sh:19]: `base_config_file` should be localized.
```suggestion
  local base_config_file
  for base_config_file in ~/.ssh/config /etc/ssh/ssh_config; do
```

The same for the variable `known_hosts_file` used later.

File: completions/ssh.completion.sh [hunk 2]
```diff
@@ -16,37 +16,42 @@ function _omb_completion_ssh {
...
+    # parse all defined hosts from config file
+    if [[ -r $base_config_file ]]; then
+      local basedir
+      basedir=$(dirname $base_config_file)
+      local -a config_files=("$base_config_file")
+
+      # check if config file contains Include options
+      local -a include_patterns
+      _omb_util_split include_patterns "$(awk -F' ' '/^Include/{print $2}' ${base_config_file} 2>/dev/null)" $'\n'
```
Comment (akinomyoga) [inline completions/ssh.completion.sh:28]:
```suggestion
      _omb_util_split include_patterns "$(awk -F' ' '/^Include/{print $2}' "${base_config_file}" 2>/dev/null)" $'\n'
```

The parameter expansion should be properly quoted.

File: completions/ssh.completion.sh [hunk 3]
```diff
@@ -16,37 +16,42 @@ function _omb_completion_ssh {
...
-      [[ -s $include_file ]] && config_files+=("$include_file")
-    done
-
-    COMPREPLY+=($(compgen -W "$(awk '/^Host/ {for (i=2; i<=NF; i++) print $i}' "${config_files[@]}")" "${options[@]}"))
-  fi
+  for base_config_file in ~/.ssh/config /etc/ssh/ssh_config; do
+    # parse all defined hosts from config file
+    if [[ -r $base_config_file ]]; then
+      local basedir
+      basedir=$(dirname $base_config_file)
```
Comment (akinomyoga) [inline completions/ssh.completion.sh:23]:
```suggestion
      basedir=$(dirname "$base_config_file")
```

Actually, we want to reduce the number of forks and execs for performance. We know that these paths contain directory names separated by a slash, so we can simply use

```suggestion
      basedir=${base_config_file%/*}
```

File: completions/ssh.completion.sh [hunk 4]
```diff
@@ -16,37 +16,42 @@ function _omb_completion_ssh {
...
+      for include_file in "${include_files[@]}";do
+        # parse all defined hosts from that file
+        [[ -s $include_file ]] && config_files+=("$include_file")
+      done
 
-  # parse all hosts found in .ssh/known_hosts
-  if [[ -r ~/.ssh/known_hosts ]]; then
-    if grep -v -q -e '^ ssh-rsa' ~/.ssh/known_hosts; then
-      COMPREPLY+=($(compgen -W "$( awk '{print $1}' ~/.ssh/known_hosts | grep -v ^\| | cut -d, -f 1 | sed -e 's/\[//g' | sed -e 's/\]//g' | cut -d: -f1 | grep -v ssh-rsa)" "${options[@]}"))
+      COMPREPLY+=($(compgen -W "$(awk '/^Host/ {for (i=2; i<=NF; i++) print $i}' "${config_files[@]}")" "${options[@]}"))
```
Comment (akinomyoga) [inline completions/ssh.completion.sh:44]: We want to reduce the number of forks and execs, so we can run this outside the loop.

File: completions/ssh.completion.sh (hunk 1 above)
Comment (akinomyoga) [reply completions/ssh.completion.sh:21]: 7958763

File: completions/ssh.completion.sh (hunk 2 above)
Comment (akinomyoga) [reply completions/ssh.completion.sh]: 99a2f8c

File: completions/ssh.completion.sh (hunk 3 above)
Comment (akinomyoga) [reply completions/ssh.completion.sh]: 2fb782d

File: completions/ssh.completion.sh (hunk 4 above)
Comment (akinomyoga) [reply completions/ssh.completion.sh]: dc60e97

**Review by akinomyoga — APPROVED** [review]

**Pull request merged**